### ISSUE TITLE
Automatic operation of the ringlight

### ISSUE THREAD

**Issue by sokol07** [OpenScanEu/OpenScan2#19]
I think it would be nice to have an option for the ringlight to turn off automatically after the job is finished. What I mean is - if you leave the scanner working at night it won't keep the ringlight turned on forever but it will turn it off as soon as the job is finished (and turn it on as soon as a new job is started).
To implement this I have added a few new nodes:
![image](https://user-images.githubusercontent.com/39523480/157644788-3811ade3-6c59-4c2a-8a01-e1459b6f4b96.png)

<details>
  <summary>Here is the json of the above blocks</summary>

```
[
    {
        "id": "8ebd1dcb5db156ed",
        "type": "ui_text",
        "z": "1613373abaf77a2c",
        "group": "7aaf184330605300",
        "order": 2,
        "width": 6,
        "height": 1,
        "name": "",
        "label": "Current Status:",
        "format": "<font color= {{msg.color}} > {{msg.payload}} </font>",
        "layout": "row-spread",
        "className": "",
        "x": 320,
        "y": 160,
        "wires": []
    },
    {
        "id": "94a7aec739f9266b",
        "type": "function",
        "z": "1613373abaf77a2c",
        "name": "loadS",
        "func": "var file = 'status_internal_cam'\nlet fs = global.get('fs');\nvar filepath = '/home/pi/OpenScan/settings/';\nconst data = fs.readFileSync(filepath+file, 'utf8');\nmsg.payload = String(data);\n\nif (data === 'no camera found'){\n    msg.color = 'red'\n}\n\nreturn msg\n\n",
        "outputs": 1,
        "noerr": 0,
        "initialize": "",
        "finalize": "",
        "libs": [],
        "x": 170,
        "y": 160,
        "wires": [
            [
                "8ebd1dcb5db156ed",
                "b87aecd9dc7f94b1"
            ]
        ]
    },
    {
        "id": "2415272f42ce468c",
        "type": "link in",
        "z": "1613373abaf77a2c",
        "name": "start status",
        "links": [
            "6c6ef2255a7d39e5"
        ],
        "x": 55,
        "y": 160,
        "wires": [
            [
                "94a7aec739f9266b"
            ]
        ]
    },
    {
        "id": "0ad3d8cfcb22472b",
        "type": "ui_switch",
        "z": "1613373abaf77a2c",
        "name": "",
        "label": "Ringlight test",
        "tooltip": "",
        "group": "7aaf184330605300",
        "order": 13,
        "width": 0,
        "height": 0,
        "passthru": true,
        "decouple": "false",
        "topic": "topic",
        "topicType": "msg",
        "style": "",
        "onvalue": "1",
        "onvalueType": "num",
        "onicon": "",
        "oncolor": "",
        "offvalue": "0",
        "offvalueType": "num",
        "officon": "",
        "offcolor": "",
        "animate": false,
        "className": "",
        "x": 770,
        "y": 260,
        "wires": [
            [
                "863f97875fd752de"
            ]
        ]
    },
    {
        "id": "b87aecd9dc7f94b1",
        "type": "switch",
        "z": "1613373abaf77a2c",
        "name": "",
        "property": "payload",
        "propertyType": "msg",
        "rules": [
            {
                "t": "eq",
                "v": "--READY--",
                "vt": "str"
            },
            {
                "t": "neq",
                "v": "--READY--",
                "vt": "str"
            }
        ],
        "checkall": "true",
        "repair": false,
        "outputs": 2,
        "x": 290,
        "y": 200,
        "wires": [
            [
                "8c130831504887e3"
            ],
            [
                "21cdfb2315d39b6c",
                "a6b9dba90e2abab7"
            ]
        ]
    },
    {
        "id": "21cdfb2315d39b6c",
        "type": "change",
        "z": "1613373abaf77a2c",
        "name": "",
        "rules": [
            {
                "t": "set",
                "p": "ringlight",
                "pt": "flow",
                "to": "ringlightSlider",
                "tot": "flow"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 560,
        "y": 220,
        "wires": [
            []
        ]
    },
    {
        "id": "8c130831504887e3",
        "type": "change",
        "z": "1613373abaf77a2c",
        "name": "",
        "rules": [
            {
                "t": "set",
                "p": "ringlight",
                "pt": "flow",
                "to": "ringlightSliderTest",
                "tot": "flow"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 560,
        "y": 180,
        "wires": [
            []
        ]
    },
    {
        "id": "863f97875fd752de",
        "type": "change",
        "z": "1613373abaf77a2c",
        "name": "",
        "rules": [
            {
                "t": "set",
                "p": "ringlightSwitch",
                "pt": "flow",
                "to": "payload",
                "tot": "msg"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 980,
        "y": 260,
        "wires": [
            []
        ]
    },
    {
        "id": "a6b9dba90e2abab7",
        "type": "function",
        "z": "1613373abaf77a2c",
        "name": "",
        "func": "msg.payload = 0\nreturn msg;",
        "outputs": 1,
        "noerr": 0,
        "initialize": "",
        "finalize": "",
        "libs": [],
        "x": 540,
        "y": 260,
        "wires": [
            [
                "0ad3d8cfcb22472b"
            ]
        ]
    },
    {
        "id": "7aaf184330605300",
        "type": "ui_group",
        "name": "Settings",
        "tab": "e23b837a9f040895",
        "order": 1,
        "disp": false,
        "width": "6",
        "collapse": false,
        "className": ""
    },
    {
        "id": "e23b837a9f040895",
        "type": "ui_tab",
        "name": "Scan",
        "icon": "dashboard",
        "order": 2,
        "disabled": false,
        "hidden": false
    }
]
```
</details>

and also:
![image](https://user-images.githubusercontent.com/39523480/157645291-7e926c08-861d-4784-88ec-157ec6f79724.png)

<details>
  <summary>Here is the json code</summary>

```
[
    {
        "id": "4970d029a12015f6",
        "type": "python3-function",
        "z": "1613373abaf77a2c",
        "name": "LED",
        "func": "from OpenScan import ringlight\ntry:\n    val = msg['payload']\nexcept:\n    val = 0\n\nif val == 0:\n    ringlight(1,False)\n    ringlight(2,False)\nelif val == 1:\n    ringlight(1,False)\n    ringlight(2,True)\nelif val == 2:\n    ringlight(1,True)\n    ringlight(2,False)\nelif val == 3:\n    ringlight(1,True)\n    ringlight(2,True)\n",
        "outputs": 1,
        "x": 390,
        "y": 920,
        "wires": [
            [
                "6a33bdfe5ee85fb9"
            ]
        ]
    },
    {
        "id": "6a33bdfe5ee85fb9",
        "type": "ui_text",
        "z": "1613373abaf77a2c",
        "group": "7aaf184330605300",
        "order": 11,
        "width": 3,
        "height": 1,
        "name": "ringlight",
        "label": "Ringlight",
        "format": "",
        "layout": "row-left",
        "className": "",
        "x": 560,
        "y": 920,
        "wires": []
    },
    {
        "id": "733d276024f1d8d8",
        "type": "change",
        "z": "1613373abaf77a2c",
        "name": "",
        "rules": [
            {
                "t": "set",
                "p": "payload",
                "pt": "msg",
                "to": "ringlight",
                "tot": "flow"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 220,
        "y": 920,
        "wires": [
            [
                "4970d029a12015f6"
            ]
        ]
    },
    {
        "id": "830b9b27e1ad18ce",
        "type": "link in",
        "z": "1613373abaf77a2c",
        "name": "",
        "links": [
            "6c6ef2255a7d39e5"
        ],
        "x": 55,
        "y": 800,
        "wires": [
            [
                "733d276024f1d8d8",
                "7a7047582588f864"
            ]
        ]
    },
    {
        "id": "7a7047582588f864",
        "type": "switch",
        "z": "1613373abaf77a2c",
        "name": "",
        "property": "ringlightSwitch",
        "propertyType": "flow",
        "rules": [
            {
                "t": "eq",
                "v": "0",
                "vt": "num"
            },
            {
                "t": "eq",
                "v": "1",
                "vt": "num"
            }
        ],
        "checkall": "true",
        "repair": false,
        "outputs": 2,
        "x": 190,
        "y": 840,
        "wires": [
            [
                "817035da9995ec07"
            ],
            [
                "3520bcf2c4b51af9"
            ]
        ]
    },
    {
        "id": "817035da9995ec07",
        "type": "change",
        "z": "1613373abaf77a2c",
        "name": "",
        "rules": [
            {
                "t": "set",
                "p": "ringlightSliderTest",
                "pt": "flow",
                "to": "0",
                "tot": "num"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 390,
        "y": 800,
        "wires": [
            []
        ]
    },
    {
        "id": "3520bcf2c4b51af9",
        "type": "change",
        "z": "1613373abaf77a2c",
        "name": "",
        "rules": [
            {
                "t": "set",
                "p": "ringlightSliderTest",
                "pt": "flow",
                "to": "ringlightSlider",
                "tot": "flow"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 390,
        "y": 860,
        "wires": [
            []
        ]
    },
    {
        "id": "4dd7285c2b0fd79b",
        "type": "ui_slider",
        "z": "1613373abaf77a2c",
        "name": "ringlight",
        "label": "",
        "tooltip": "",
        "group": "7aaf184330605300",
        "order": 12,
        "width": 3,
        "height": 1,
        "passthru": true,
        "outs": "all",
        "topic": "",
        "topicType": "str",
        "min": 0,
        "max": "3",
        "step": 1,
        "className": "",
        "x": 320,
        "y": 700,
        "wires": [
            [
                "207caa543304bd9f"
            ]
        ]
    },
    {
        "id": "207caa543304bd9f",
        "type": "change",
        "z": "1613373abaf77a2c",
        "name": "",
        "rules": [
            {
                "t": "set",
                "p": "ringlightSlider",
                "pt": "flow",
                "to": "payload",
                "tot": "msg"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 560,
        "y": 700,
        "wires": [
            []
        ]
    },
    {
        "id": "7aaf184330605300",
        "type": "ui_group",
        "name": "Settings",
        "tab": "e23b837a9f040895",
        "order": 1,
        "disp": false,
        "width": "6",
        "collapse": false,
        "className": ""
    },
    {
        "id": "e23b837a9f040895",
        "type": "ui_tab",
        "name": "Scan",
        "icon": "dashboard",
        "order": 2,
        "disabled": false,
        "hidden": false
    }
]
```
</details>

There is a new button on the Scan page: "ringlight test" - switching it will turn on the ringlight even if the scanner is in "ready" state. The button will be turned off when a job is started so the ringlight will turn off after the job is finished. The set value of ringlight slider remains unchanged.

As the changes are a bit extensive I decided to fork OS to my github and I'll create a pull request soon, which will contain solution to this issue and also #16 and #17

**Comment by sokol07** [comment]
There is an error in this solution, at the start of the flow there is no message for the LED function and everything crashes. I'm working on a fix.

**Comment by sokol07** [comment]
I fixed the error, changed the code in the forked version and in the json code above.

**Comment by OpenScanEu** [comment]
I have used a little different approach, where I added a idle-timer, which can be called through http://127.0.0.1:1312/ping which will return a time in seconds.
The timer will be reset to zero whenever the motor is being moved or the ringlight status changes. I set a timeout value to 300s, which can be modified by changing the file /home/pi/OpenScan/settings/timeout_ringlight
I intend to use the same timer (or a very similar approach) to turn off the stepper motors (which is currently not possible due to the hardware, but a new version should be out at some point)

**Comment by sokol07** [comment]
How does it behave when the device is idling? If I leave the device for some time, return to it later and ai want to "wake up" the ring light I have to move the table or change the ring light switch position?
I have a feeling that's now intuitive, maybe it could be woken up by any action from UI?
My second concern is the timeout mechanism - if taking the picture takes longer time (because, for example, you're using focus stacking) you have to set a long timeout if i understand correctly. This way the ring light will also stay lit for long time after the job is done.
I don't say that's wrong approach and that it doesn't make sense, however I liked the way my solution worked, turning the ring light on/off as soon as the job was started or finished (I think the motors could be operated in future in the same way) - is there a reason you decided to switch to the timeout mechanism?

**Comment by OpenScanEu** [comment]
> How does it behave when the device is idling? If I leave the device for some time, return to it later and ai want to "wake up" the ring light I have to move the table or change the ring light switch position?
The device won't wake up automatically, and you will have to use the ring light switch again

> My second concern is the timeout mechanism - if taking the picture takes longer time (because, for example, you're using focus stacking) you have to set a long timeout if i understand correctly. This way the ring light will also stay lit for long time after the job is done.
Yes, this is right. For this reason, I have set the timeout to a large value of 300s. So the ring light will stay on for 5 mins after the routine is done.

I have tried your solution, and it is much more efficient and elegant. But it somehow conflicted with my habit/typical workflow:

I turn on the device, turn on a ring light, spray the object and place it on the turntable, adjust the focus/brightness and start the routine. After a couple of minutes, I return to the device, change the object's orientation and just hit run to capture a second set. In the meantime, I prepare a second scan object.

So, long story short, I switch off the ring light only when powering off the device. So for me the timeout mechanism helps, that I do not forget the ring light (and in the future: motors), when I leave my place or forget to turn off the device. And I am too lazy to turn on the ring light again after each routine, when I need to adjust the brightness level for a new object.

**Comment by sokol07** [comment]
Ok, I get it!

**Comment by sokol07** [comment]
Ok, I get it!
Considering your workflow I understand that the timeout mechanism can be more useful.
However, maybe we could meet in the middle with the solution - maybe the timeout mechanism can be active only when status is "--READY--" so that setting the timeout to, for example, 1 second would turn the light off faster when the job is completed?

**Comment by sokol07** [comment]
@OpenScanEu can we pull this value out from the config files to the settings page?

**Comment by OpenScanEu** [comment]
@sokol07 Absolutely, I will do so in the update coming next week (only minor changes so far)
